chia/full_node/block_store.py

```diff
@@ -60,11 +60,14 @@ async def create(cls, db_wrapper: DBWrapper):
             )
 
             # Height index so we can look up in order of height for sync purposes
-            await self.db.execute("CREATE INDEX IF NOT EXISTS full_block_height on full_blocks(height)")
+            await self.db.execute("CREATE INDEX IF NOT EXISTS height on full_blocks(height)")
             await self.db.execute(
-                "CREATE INDEX IF NOT EXISTS is_fully_compactified on full_blocks(is_fully_compactified)"
+                "CREATE INDEX IF NOT EXISTS is_fully_compactified ON"
+                " full_blocks(is_fully_compactified, in_main_chain) WHERE in_main_chain=1"
+            )
+            await self.db.execute(
+                "CREATE INDEX IF NOT EXISTS main_chain ON full_blocks(height, in_main_chain) WHERE in_main_chain=1"
```

Comment:
All three indexes (height, is_fully_compactified, main_chain) are created on columns for similar purpose.

It may make sense to only create one index and create basic statistics (`analyze` for `sqlite_stat1`) for the indexes to be able to use [index skip scans](https://www.sqlite.org/optoverview.html#skipscan).
This should also slightly reduce the db size and reduce index maintainence overhead and may help work around the problem that SQLite uses [at most one index](https://www.sqlite.org/optoverview.html#choosing_between_multiple_indexes).

I'd try index order `ON full_blocks(height, in_main_chain, is_fully_compactified) WHERE in_main_chain=1` and hint the statements where index isn't used properly with `INDEXED BY` and [+](https://www.sqlite.org/optoverview.html#disqualifying_where_clause_terms_using_unary_)

Reply:
would it require running `analyze` periodically though? because I'm not sure that's feasible

Reply:
if I add `height` to the index, the benchmark slows down a *lot*. inserting the blocks takes twice as long.. I'm not sure why

Reply:
periodic analyze wouldn't be necessary in my opinion, because the distribution won't change significantly. 

for instance, this is how the stats look like on my production full node:
```
sqlite> select * from sqlite_stat1;
+-----------------------+------------------------------------------+----------------+
|          tbl          |                   idx                    |      stat      |
+-----------------------+------------------------------------------+----------------+
| full_blocks           | is_fully_compactified                    | 1291224 645612 |
| full_blocks           | full_block_height                        | 1291224 2      |
| full_blocks           | sqlite_autoindex_full_blocks_1           | 1291224 1      |
| block_records         | height                                   | 1291224 2      |
| block_records         | peak                                     | 1 1            |
| block_records         | sqlite_autoindex_block_records_1         | 1291224 1      |
| sub_epoch_segments_v3 | sqlite_autoindex_sub_epoch_segments_v3_1 | 3356 1         |
| coin_record           | coin_spent_index                         | 33884059 92    |
| coin_record           | coin_parent_index                        | 33884059 5     |
| coin_record           | coin_puzzle_hash                         | 33884059 8     |
| coin_record           | coin_confirmed_index                     | 33884059 75    |
| coin_record           | sqlite_autoindex_coin_record_1           | 33884059 1     |
+-----------------------+------------------------------------------+----------------+
```

Reply:
it depends on how you want to set it up.

As we all share the same databases, the statistics will be similar for every node and 
you could update the stats as part of a new sw-release manually. 
I'd guess the data distribution/cardinality won't change out of a sudden.

[manually set statistics](https://www.sqlite.org/optoverview.html#manctrl)
[sqlite_stat1_table](https://www.sqlite.org/fileformat.html#the_sqlite_stat1_table)

From a db performance perspective I'd even use more statistics, but I noticed that (at least on my side) sqlite isn't compiled with sqlite_enable_stat3 or sqlite_enable_stat4.

Reply:
I would like to punt this to a separate PR. I tried a few simple approaches, but it's not trivial, especially to work in v1 schema mode as well